File: algorithms/java/CountingSort.java
import java.util.*;

public class CountingSort
{
	public static void sort(Integer[] array)
	{
		int min = Collections.min(Arrays.asList(array));
		int max = Collections.max(Arrays.asList(array));
		int range = (max - min) + 1;
		int[] count = new int[range];
		int[] output = new int[array.length];

		for(int i = 0; i < array.length; i++) count[array[i] - min]++;
		for(int i = 1; i < count.length; i++) count[i] += count[i - 1];

		for(int i = array.length - 1; i >= 0; i--)
		{
			output[count[array[i] - min] - 1] = array[i];
			count[array[i] - min]--;
		}

		for(int i = 0; i < array.length; i++)
		{
			array[i] = output[i];
		}
	}
}

File: algorithms/go/counting_sort.go
package algorithms

func CountingSort(slice []int) {
	min := slice[0]
	max := slice[0]

	for i := 0; i < len(slice); i++ {
		if slice[i] > max {
			max = slice[i]
		}

		if slice[i] < min {
			min = slice[i]
		}
	}

	section := (max - min) + 1
	count := make([]int, section)
	sortedSlice := make([]int, len(slice))

	for i := 0; i < len(slice); i++ {
		count[slice[i] - min]++
	}

	for i := 1; i < len(count); i++ {
		count[i] += count[i - 1]
	}

	for i := len(slice) - 1; i >= 0; i-- {
		sortedSlice[count[slice[i] - min] - 1] = slice[i]
		count[slice[i] - min]--
	}

	copy(slice, sortedSlice)
}